test/integration/formatting_stack/strategies.clj

```diff
@@ -0,0 +1,153 @@
+
+(defn expect-sane-output! [corpus]
+  (let [n (count corpus)]
+    (is (pos? n))
+    (doseq [^String filename corpus
+            :let [f (File. filename)]]
+      (is (-> f .exists))
+      (testing "It emits absolutized filenames"
+        (is (= filename
+               (-> f .getCanonicalPath)))))))
+
+(deftest all-files
+
+  (assert-pristine-git-status!)
+
+  (testing "It runs without errors, exercising its specs, even in face of deleted (but not staged) files"
+    (try
+      (-> deletable-file .delete)
+      (expect-sane-output! (sut/all-files :files []))
+      (finally
+        (sh "git" "checkout" (str deletable-file))
+        (assert (-> deletable-file .exists)))))
+
+  (assert-pristine-git-status!)
+
+  (testing "It runs without errors, exercising its specs, even in face of files staged for deletion"
+    (try
+      (-> deletable-file .delete)
+      (sh "git" "add" "-A")
+      (expect-sane-output! (sut/all-files :files []))
+      (finally
+        (sh "git" "reset" "--" (str deletable-file))
+        (sh "git" "checkout" (str deletable-file))
+        (assert (-> deletable-file .exists))))))
+
+(deftest git-diff-against-default-branch
+
+  (assert-pristine-git-status!)
+
+  (testing "It runs without errors, exercising its specs, even in face of deleted (but not staged) files"
+    (let [;; f-stack's first ever commit. This ensures a large, diverse corpus:
+          first-commit "ebb0ca8"]
+
+      (try
+        (-> deletable-file .delete)
+        (expect-sane-output! (sut/git-diff-against-default-branch :target-branch first-commit))
+        (finally
+          (sh "git" "checkout" (str deletable-file))
+          (assert (-> deletable-file .exists))))))
+
+  (assert-pristine-git-status!)
+
+  (testing "It runs without errors, exercising its specs, even in face of files staged for deletion"
+    (let [;; f-stack's first ever commit. This ensures a large, diverse corpus:
+          first-commit "ebb0ca8"]
+
+      (try
+        (-> deletable-file .delete)
+        (sh "git" "add" "-A")
+        (expect-sane-output! (sut/git-diff-against-default-branch :target-branch first-commit))
+        (finally
+          (sh "git" "reset" "--" (str deletable-file))
+          (sh "git" "checkout" (str deletable-file))
+          (assert (-> deletable-file .exists)))))))
+
+(deftest git-not-completely-staged
+
+  (assert-pristine-git-status!)
+
+  (testing "Non-completely-added files show up"
+    (try
+      (spit createable-filename creatable-contents)
+
+      (is (= [(-> createable-filename File. .getCanonicalPath)]
+             (sut/git-not-completely-staged :files [])))
+
+      (testing "Added files don't show up"
+        (try
+          (sh "git" "add" "-A")
+          (is (= []
+                 (sut/git-not-completely-staged :files [])))
+          (finally
+            (sh "git" "reset" "."))))
+
+      (finally
+        (-> createable-filename File. .delete))))
```

Comment:
exceptions will be caught inside of the `is`, right? does this work how the way its intended?

Reply:
> exceptions will be caught inside of the is, right?

`finally` runs 100% unconditionally, whether an exception was thrown or not.